docs/jetting.md

```diff
@@ -0,0 +1,709 @@
+want to get data out of your atoms.
+
+If you know that the data fits in 32 bits or less, you can use
+```
+   u3r_word(c3_w    a_w, u3_atom b);
+```
+
+If it is longer than 32 bits, use
+```
+  u3r_words(c3_w    a_w, c3_w    b_w, c3_w*   c_w, u3_atom d);
+```
+or
+```
+   u3r_bytes(c3_w    a_w, c3_w    b_w, c3_y*   c_y, u3_atom d)
+```
+
+(N.B. You may want to use u3r_bytes regardless, as 
+
+The actual meat of the function is up to you.  What is it supposed to do?
+
+Now we move on to return semantics.
+
+First, you can transfer raw values into nouns using
+```
+    u3_noun u3i_words(c3_w a_w, const c3_w* b_w)
+```
+and you can build cells out of nouns using
+```
+	XXX
+```
+There are two facets here: (i) data format, (ii) memory allocation.
+
+Regarding data format, if the hoon is expected to return a single atom (e.g. if the hoon looks like this
+```
+      ++  make-k            
+        ~/  %make-k
+        |=  [has=@uvI prv=@]     ::  <---- input parguments
+        ^-  @                    ::  <---- return value is a single value of type '@' (atom)
+		...
+```
+then your C code - at least when you're stubbing it out - can do something like
+```
+	 return(123);
+```
+Or, if you want to create an atom more formally, you can build it like this
+```
+     // this variable is on the stack and will disappear
+	 unsigned char nonce32[32];  
+
+	 // this allocates an indirect (> 31 bits) atom in the loom, 
+	 // does appropriate reference count, and returns the 32 bit handle to the atom
+     u3_noun nonce = u3i_words(8, (const c3_w*) nonce32);
+
+	 // this returns the 32 bit handle to the atom
+	 return(nonce);
+```
+If, on the other hand, your hoon looks like
+```
+      ++  ecdsa-raw-sign                                ::  generate signature
+        ~/  %ecdsa-raw-sign
+        |=  [has=@uvI prv=@]
+        ^-  [v=@ r=@ s=@]
+        ...
+```
+ending your C code with
+```
+	 return(123);
+```
+is wrong and will result in a dojo error because you are returning a single atom, instead of a list of three atoms.
+
+Instead do one of these:
+```
+     return(u3nc(a, b));		// for two atoms 
+     return(u3nt(a, b, c));		// for three atoms 
+     return(u3nq(a, b, c, d));	// for four atoms 
+```
+If you need to return a longer list, you can compose your own.  Look
+at the definitions of these three functions and you will see that they
+are just recursive calls to the cell constructor u3i_cell() e.g.
+```
+		 u3i_cell(a, u3i_cell(b, u3i_cell(c, d));
+```
+
+XXX more here about memory model, allocation, ownership
+
+
+
+## Compile the C code
+
+You've edited C code; now you need to compile it to build a new urbit executable!
+
+
+1. cd ~/tlon/urbit
+1. rm -rf build
+1. configure: choose either
+    1. meson configure -Dbuildtype=debug build
+    1. meson configure -Dbuildtype=release build
+	Note that debug builds run much much slower.  Booting a debug builkd of urbit may take several minutes.
```

Comment:
You can use `-Dbuildtype=debugoptimized` to get a binary with debug symbols that will still run reasonably quickly (I think it's `-O2`, while `release` is `-O3`).

Reply:
added